lib/Target/SubstraitPB/Import.cpp

```diff
@@ -0,0 +1,101 @@
+
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/OwningOpRef.h"
+#include "structured/Dialect/Substrait/IR/Substrait.h"
+
+#include <google/protobuf/text_format.h>
+#include <substrait/plan.pb.h>
+
+using namespace mlir;
+using namespace mlir::substrait;
+
+namespace pb = google::protobuf;
+
+namespace {
+
+// Forward declaration for the import function of the given message type.
+//
+// We need one such function for most message types that we want to import. The
+// forward declarations are necessary such all import functions are available
+// for the definitions indepedently of the order of these definitions. The
+// message type passed to the function (specified by `MESSAGE_TYPE`) may be
+// different than the one it is responsible for: often the target op type
+// (specified by `OP_TYPE`) depends on a nested field value (such as `oneof`)
+// but the import logic needs the whole context; the message that is passed in
+// is the most deeply nested message that provides the whole context.
+#define DECLARE_IMPORT_FUNC(MESSAGE_TYPE, ARG_TYPE, OP_TYPE)                   \
+  static FailureOr<OP_TYPE> import##MESSAGE_TYPE(                              \
+      ImplicitLocOpBuilder builder, const ::substrait::ARG_TYPE &message);
+
+DECLARE_IMPORT_FUNC(Plan, Plan, PlanOp)
+DECLARE_IMPORT_FUNC(PlanRel, PlanRel, PlanRelOp)
+
+static FailureOr<PlanOp> importPlan(ImplicitLocOpBuilder builder,
+                                    const ::substrait::Plan &message) {
+  const ::substrait::Version &version = message.version();
+  auto planOp = builder.create<PlanOp>(
+      version.major_number(), version.minor_number(), version.patch_number(),
+      version.git_hash(), version.producer());
+  planOp.getBody().push_back(new Block());
+
+  for (int i = 0; i < message.relations_size(); i++) {
+    OpBuilder::InsertionGuard insertGuard(builder);
+    builder.setInsertionPointToEnd(&planOp.getBody().front());
+    if (failed(importPlanRel(builder, message.relations(i))))
+      return failure();
+  }
+
+  return planOp;
+}
+
+static FailureOr<PlanRelOp> importPlanRel(ImplicitLocOpBuilder builder,
+                                          const ::substrait::PlanRel &message) {
+  switch (message.rel_type_case()) {
+  case ::substrait::PlanRel::RelTypeCase::kRel: {
+    auto planRelOp = builder.create<PlanRelOp>();
+    // TODO(ingomueller): import content once defined.
+    return planRelOp;
+  }
+  default:
+    return failure();
+  }
+}
+
+} // namespace
+
+namespace mlir {
+namespace substrait {
+
+OwningOpRef<ModuleOp> translateProtobufToSubstrait(llvm::StringRef input,
+                                                   MLIRContext *context) {
+  auto plan = std::make_unique<::substrait::Plan>();
+  if (!pb::TextFormat::ParseFromString(input.str(), plan.get()))
+    return {};
+
+  // XXX: Is this really the right way/place to do this?
+  context->loadDialect<SubstraitDialect>();
```

Comment:
You probably want to do it in the translate registration yes (its not bad to do it here as you don't have the same "dependent dialect" facility that passes have)

Reply:
Thanks for the input! I looked how the other translations do it: both [LLVM](https://github.com/llvm/llvm-project/blob/6b62a9135a/mlir/lib/Target/LLVMIR/ModuleImport.cpp#L2079) and [SPIRV](https://github.com/llvm/llvm-project/blob/6b62a9135a/mlir/lib/Target/SPIRV/TranslateRegistration.cpp#L41) load their respective dialect in their "to MLIR" translate function explicitly.

Note that I already use the [corresponding hook](https://github.com/llvm/llvm-project/blob/6b62a9135a/mlir/include/mlir/Tools/mlir-translate/Translation.h#L120-L121) for *registering* the dialects (namely with `registerSubstraitDialects` in `structured-translate.cpp`).

In the latest commit, I simply dropped the comment.